### ISSUE TITLE
concepts appear to behave incorrectly in conjunction with 'var'

### ISSUE THREAD

**Issue by ghost** [nim-lang/Nim#1033]
``` nimrod
proc put*[T](c: seq[T], x: T) =
  discard

when true:
  proc put*[T](c: var seq[T], x: T) =
    c.add x

when true:
  type
    Container[T] = generic c
      put(c, T)
else:
  type
    Container[T] = generic c
      put(var c, T) # 'var' here doesn't seem to have any effect.

var x = @[1]

# These pass, but only without the 'put' overloads.
assert(seq[int] is Container[int])
assert(x is Container[int])

# These fail. Should they?
assert(seq is Container)
assert(x is Container)
```

**Comment by andreaferretti** [comment]
The followgin example

```nim
proc put*[T](c: seq[T], x: T) =
  discard

proc put*[T](c: var seq[T], x: T) =
  discard

type
  Container[T] = concept c
    put(c, T)

var x = @[1]

# These pass, but only without the 'put' overloads.
echo(seq[int] is Container[int])
echo(x is Container[int])
echo(seq is Container)
echo(x is Container)
```
as of today prints

```
true
true
false
true
```

Arguably, these are the right answers. Should this be closed?

**Comment by zah** [comment]
This is a bit complicated case to support, but I think `seq is Container` should be true as well.

**Comment by andreaferretti** [comment]
It's complicated. Essentially one has to decide on the order of quantifiers. The first two are check a concrete type against a non generic concept, so there is little doubt. The latter one is again checking a concrete type, this time against a generic concept - ok.

The third one is subtle. On the one hand `seq[A] is Container[A]` should be true, for each concrete type `A` - in fact we see this for `int` the first example. Hence `seq[A] is Container` is also true for every concrete `A`. One may want to generalize and say that `seq is Container` is also true.

On the other hand, `seq is Container[A]` is false for every choice of `A` - there is no concrete `A` for which `seq` is a `Container[A]`. Hence one may want to conclude that `seq is Container` is false.

There is no objective right answer here, one just has to choose a convention. This is why I was telling that answer three was ok

**Comment by zah** [comment]
The compiler has a certain notion of ordering of generic types. 

If the set of types conforming to a given generic signature is a strict subset of the types conforming to another generic signature, we'll consider the first generic signature "more specific" for overloading resolution purposes (this is planned to arrive soon). 

The `is` operator and the overload checks within the concept bodies can use the same "more specific" definition when comparing types against each other. Under these rules, `seq is Container` will end up comparing `auto` against `auto` (these are the unconstrained `T` params), which is considered the same set of types (and thus the result will be true).

**Comment by ephja** [comment]
@andreaferretti 'var' in the concept specification still doesn't work though. Omitting 'var' however results in a match for both var and non-var prototypes. It's probably not intended, but it does enable various patterns.

**Comment by ephja** [comment]
@zah Did that patch fix the bug? It seems like it did.

**Comment by zah** [comment]
Yes, the the mode obscure case of `seq is Container` is still not working correctly.
